lib/go-atscfg/remapdotconfig.go

```diff
@@ -0,0 +1,369 @@
+	}
+
+	textLines := []string{}
+
+	for _, ds := range dses {
+		remapText := ""
+		if ds.Type == tc.DSTypeAnyMap {
+			if ds.RemapText == nil {
+				log.Errorln("ds '" + ds.Name + "' is ANY_MAP, but has no remap text - skipping")
+				continue
+			}
+			remapText = *ds.RemapText + "\n"
+			textLines = append(textLines, remapText)
+			continue
+		}
+
+		remapLines, err := MakeEdgeDSDataRemapLines(ds, server)
+		if err != nil {
+			log.Errorln("making remap lines for DS '" + ds.Name + "' - skipping! : " + err.Error())
+			continue
+		}
+
+		for _, line := range remapLines {
+			profilecacheKeyConfigParams := (map[string]string)(nil)
+			if ds.ProfileID != nil {
+				profilecacheKeyConfigParams = profilesCacheKeyConfigParams[*ds.ProfileID]
+			}
+			remapText = BuildRemapLine(cacheURLConfigParams, atsMajorVersion, server, serverPackageParamData, remapText, ds, line.From, line.To, profilecacheKeyConfigParams)
+		}
+		textLines = append(textLines, remapText)
+	}
+
+	text := header
+	sort.Sort(sort.StringSlice(textLines))
+	for _, line := range textLines {
+		text += line
+	}
+	return text
+}
+
+// BuildRemapLine builds the remap line for the given server and delivery service.
+// The cacheKeyConfigParams map may be nil, if this ds profile had no cache key config params.
+func BuildRemapLine(cacheURLConfigParams map[string]string, atsMajorVersion int, server *ServerInfo, pData map[string]string, text string, ds RemapConfigDSData, mapFrom string, mapTo string, cacheKeyConfigParams map[string]string) string {
+	// ds = 'remap' in perl
+
+	mapFrom = strings.Replace(mapFrom, `__http__`, server.HostName, -1)
+
+	if _, hasDSCPRemap := pData["dscp_remap"]; hasDSCPRemap {
+		text += "map	" + mapFrom + "     " + mapTo + ` @plugin=dscp_remap.so @pparam=` + strconv.Itoa(ds.DSCP)
+	} else {
+		text += "map	" + mapFrom + "     " + mapTo + ` @plugin=header_rewrite.so @pparam=dscp/set_dscp_` + strconv.Itoa(ds.DSCP) + ".config"
+	}
+
+	if ds.EdgeHeaderRewrite != nil && *ds.EdgeHeaderRewrite != "" {
+		text += ` @plugin=header_rewrite.so @pparam=` + EdgeHeaderRewriteConfigFileName(ds.Name)
+	}
+
+	if ds.SigningAlgorithm != nil && *ds.SigningAlgorithm != "" {
+		if *ds.SigningAlgorithm == tc.SigningAlgorithmURLSig {
+			text += ` @plugin=url_sig.so @pparam=url_sig_` + ds.Name + ".config"
+		} else if *ds.SigningAlgorithm == tc.SigningAlgorithmURISigning {
+			text += ` @plugin=uri_signing.so @pparam=uri_signing_` + ds.Name + ".config"
+		}
+	}
+
+	if ds.QStringIgnore != nil {
+		if *ds.QStringIgnore == tc.QueryStringIgnoreDropAtEdge {
+			dqsFile := "drop_qstring.config"
+			text += ` @plugin=regex_remap.so @pparam=` + dqsFile
+		} else if *ds.QStringIgnore == tc.QueryStringIgnoreIgnoreInCacheKeyAndPassUp {
+			if _, globalExists := cacheURLConfigParams["location"]; globalExists {
+				log.Debugln("qstring_ignore == 1, but global cacheurl.config param exists, so skipping remap rename config_file=cacheurl.config parameter") // TODO warn? Perl was a debug
```

Comment:
Is debug the right place to log this? This is a pretty serious configuration snafu. The DS is configured to ignore the query string, but this is ignoring the request to ignore it. I worry that a debug log will wind up ignored.

Reply:
Changed to a Warning.